File: sdk/java-holons/src/main/java/org/organicprogramming/holons/DiscoveryTypes.java
package org.organicprogramming.holons;

import java.util.ArrayList;
import java.util.List;

/** Shared types and constants for the uniform discovery/connect APIs. */
public final class DiscoveryTypes {
    public static final int LOCAL = 0;
    public static final int PROXY = 1;
    public static final int DELEGATED = 2;

    public static final int SIBLINGS = 0x01;
    public static final int CWD = 0x02;
    public static final int SOURCE = 0x04;
    public static final int BUILT = 0x08;
    public static final int INSTALLED = 0x10;
    public static final int CACHED = 0x20;
    public static final int ALL = 0x3F;

    public static final int NO_LIMIT = 0;
    public static final int NO_TIMEOUT = 0;

    private DiscoveryTypes() {
    }

    public static final class IdentityInfo {
        public String givenName = "";
        public String familyName = "";
        public String motto = "";
        public List<String> aliases = new ArrayList<>();
    }

    public static final class HolonInfo {
        public String slug = "";
        public String uuid = "";
        public IdentityInfo identity = new IdentityInfo();
        public String lang = "";
        public String runner = "";
        public String status = "";
        public String kind = "";
        public String transport = "";
        public String entrypoint = "";
        public List<String> architectures = new ArrayList<>();
        public boolean hasDist;
        public boolean hasSource;
    }

    public static final class HolonRef {
        public String url = "";
        public HolonInfo info;
        public String error = "";
    }

    public static final class DiscoverResult {
        public List<HolonRef> found = new ArrayList<>();
        public String error = "";
    }

    public static final class ResolveResult {
        public HolonRef ref;
        public String error = "";
    }

    public static final class ConnectResult {
        public Object channel;
        public String uid = "";
        public HolonRef origin;
        public String error = "";
    }
}

File: sdk/kotlin-holons/src/main/kotlin/org/organicprogramming/holons/DiscoveryTypes.kt
package org.organicprogramming.holons

const val LOCAL = 0
const val PROXY = 1
const val DELEGATED = 2

const val SIBLINGS = 0x01
const val CWD = 0x02
const val SOURCE = 0x04
const val BUILT = 0x08
const val INSTALLED = 0x10
const val CACHED = 0x20
const val ALL = 0x3F

const val NO_LIMIT = 0
const val NO_TIMEOUT = 0

data class IdentityInfo(
    val givenName: String,
    val familyName: String,
    val motto: String = "",
    val aliases: List<String> = emptyList(),
)

data class HolonInfo(
    val slug: String,
    val uuid: String,
    val identity: IdentityInfo,
    val lang: String,
    val runner: String,
    val status: String,
    val kind: String,
    val transport: String,
    val entrypoint: String,
    val architectures: List<String>,
    val hasDist: Boolean,
    val hasSource: Boolean,
)

data class HolonRef(
    val url: String,
    val info: HolonInfo? = null,
    val error: String? = null,
)

data class DiscoverResult(
    val found: List<HolonRef> = emptyList(),
    val error: String? = null,
)

data class ResolveResult(
    val ref: HolonRef? = null,
    val error: String? = null,
)

data class ConnectResult(
    val channel: Any? = null,
    val uid: String = "",
    val origin: HolonRef? = null,
    val error: String? = null,
)
